lib/cretonne/src/ir/types.rs

```diff
@@ -35,6 +35,22 @@ pub const VOID: Type = Type(0);
 // 512-bit SIMD vectors.
 include!(concat!(env!("OUT_DIR"), "/types.rs"));
 
+/// Possible derivation functions for types. Used in runtime typechecking
+pub enum DerivedFunctions {
```

Comment:
On second thought, it might make sense to add an assert before the check that the "derived" type is not None (i.e. there is no over/underflow)?